### PR TITLE
Fix crash trying to make a github issue report caused by compiler metadata being missing.

### PR DESCRIPTION
**Pull request by BillyONeal**
Resolves https://github.com/microsoft/vcpkg/issues/32770

I think there is an argument that `AbiInfo` should be 2 different types but I did enough drive-by fixes here:

* Fixed several functions that only wanted pre_build_info and/or toolset but asked for a whole AbiTag.
* Fixed compute_abi_tag requiring that abi_tag was already filled in before it could be called. (???)

### CODE REVIEW

File: src/vcpkg/commands.build.cpp [hunk 1]
```diff
@@ -1127,53 +1131,57 @@ namespace vcpkg
...
         {
             if (port_file.filename() == ".DS_Store")
             {
                 continue;
             }
-            abs_port_file = port_dir;
-            abs_port_file /= port_file;
 
+            files.push_back(port_dir / port_file);
+            const auto& abs_port_file = files.back();
```
Comment (Thomas1664) [inline src/vcpkg/commands.build.cpp:1171]: Can't you use `emplace_back` here so you already have the reference to the last element?

Comment (BillyONeal) [reply src/vcpkg/commands.build.cpp:1171]: If it works on our compiler support matrix, sure

Comment (Thomas1664) [reply src/vcpkg/commands.build.cpp:1171]: > If it works on our compiler support matrix, sure

It should work: https://en.cppreference.com/w/cpp/compiler_support/17#C.2B.2B17_library_features
Search for [P0084R2](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0084r2.pdf).

Comment (BillyONeal) [reply src/vcpkg/commands.build.cpp:1171]: Thanks :)

**Review by JavierMatosD — APPROVED** [review]

**Pull request merged**